### ISSUE TITLE
Invalid column name in ORDER BY clause using 5.1.7

### ISSUE THREAD

**Issue by bwbasheer** [rails-sqlserver/activerecord-sqlserver-adapter#757]
Eo.find('17127     ') results in a SQL EXEC statement that contains an invalid ORDER BY clause:

irb(main):005:0> Eo.find('17127     ')
  SQL (0.2ms)  USE [veprod]
  Eo Load (0.8ms)  EXEC sp_executesql N'SELECT  [Eo].* FROM [Eo] WHERE [Eo].[eoID] = @0  ORDER BY [Eo].["jurisdictionID", "eotypeID"] ASC OFFSET 0 ROWS FETCH NEXT @1 ROWS ONLY', N'@0 int, @1 int', @0 = 17127, @1 = 1  [["eoID", nil], ["LIMIT", nil]]
Traceback (most recent call last):
        1: from (irb):5
ActiveRecord::StatementInvalid (TinyTds::Error: Invalid column name '"jurisdictionID", "eotypeID"'.: EXEC sp_executesql N'SELECT  [Eo].* FROM [Eo] WHERE [Eo].[eoID] = @0  ORDER BY [Eo].["jurisdictionID", "eotypeID"] ASC OFFSET 0 ROWS FETCH NEXT @1 ROWS ONLY', N'@0 int, @1 int', @0 = 17127, @1 = 1)

Here is the Eo model
class Eo < ActiveRecord::Base
  self.table_name = "Eo"
  self.primary_key = "eoID"
  belongs_to :jurisdiction, foreign_key: "jurisdictionID"
  belongs_to :party, foreign_key: "partyID"
end

Would appreciate some guidance on how to research this issue.

**Comment by wpolicarpo** [comment]
@bwbasheer do you have a composite primary key on `Eo` table on `jurisdictionID` and `eotypeID` columns? If so, can this composite PK be converted into a simple PK on `eoID`?

**Comment by bwbasheer** [comment]
Thanks for your response.  As can be seen in the model definition above, there is no explicit composite primary key.  This works fine in earlier Rails version settings.

**Comment by wpolicarpo** [comment]
Can you post the output for the comand below when you execute in a rails console?

```ruby 
Eo.columns_hash.slice('eoID', 'jurisdictionID', 'eotypeID')
```

**Comment by bwbasheer** [comment]
Sure.  Thanks:

irb(main):033:0> Eo.columns_hash.slice('eoID', 'jurisdictionID', 'eotypeID')
=> {"eoID"=>#<ActiveRecord::ConnectionAdapters::SQLServerColumn:0x00000000089d3e58 @sqlserver_options={:ordinal_position=>1, :is_primary=>false, :is_identity=>true}, @name="eoID", @table_name="Eo", @sql_type_metadata=#<ActiveRecord::ConnectionAdapters::SQLServer::SqlTypeMetadata:0x0000000008a2c148 @sqlserver_options={:sqlserver_options=>{:ordinal_position=>1, :is_primary=>false, :is_identity=>true}}, @sql_type="int(4)", @type=:integer, @limit=4, @precision=nil, @scale=nil>, @null=false, @default=nil, @default_function=nil, @collation=nil, @comment=nil>, "jurisdictionID"=>#<ActiveRecord::ConnectionAdapters::SQLServerColumn:0x00000000089d2f30 @sqlserver_options={:ordinal_position=>2, :is_primary=>true, :is_identity=>false}, @name="jurisdictionID", @table_name="Eo", @sql_type_metadata=#<ActiveRecord::ConnectionAdapters::SQLServer::SqlTypeMetadata:0x00000000089d30c0 @sqlserver_options={:sqlserver_options=>{:ordinal_position=>2, :is_primary=>true, :is_identity=>false}}, @sql_type="char(10)", @type=:char, @limit=10, @precision=nil, @scale=nil>, @null=false, @default=nil, @default_function=nil, @collation="SQL_Latin1_General_CP1_CI_AS", @comment=nil>, "eotypeID"=>#<ActiveRecord::ConnectionAdapters::SQLServerColumn:0x00000000089d2b98 @sqlserver_options={:ordinal_position=>3, :is_primary=>true, :is_identity=>false}, @name="eotypeID", @table_name="Eo", @sql_type_metadata=#<ActiveRecord::ConnectionAdapters::SQLServer::SqlTypeMetadata:0x00000000089d2d50 @sqlserver_options={:sqlserver_options=>{:ordinal_position=>3, :is_primary=>true, :is_identity=>false}}, @sql_type="int(4)", @type=:integer, @limit=4, @precision=nil, @scale=nil>, @null=false, @default=nil, @default_function=nil, @collation=nil, @comment=nil>}

**Comment by wpolicarpo** [comment]
As you can see per the output, `eoID` **is not** the PK of your table, but you have a composite PK on (`jurisdictionID`, `eotypeID`). Look `is_primary` attribute for those columns.

This might be considered a bug in the adapter but since Rails does not support composite PKs I asked you if you could fix your table to reflect what you have on your model.

PS: Please when posting code/output please consider using the proper code formatting on Github to make it easier for us to help you.

**Comment by bwbasheer** [comment]
Thank you for the information and suggestions.

**Comment by bwbasheer** [comment]
Wanderson: Given that the model declaration defines eoID as the primary key, can you say what causes the adapter to treat it as part of a composite key?

**Comment by wpolicarpo** [comment]
When you're fetching a record from the database, the adapter will try to make your query deterministic, shortly explaining, your query should always return the same result if data hasn't chanegd.

It does that by calling [`make_Fetch_Possible_And_Deterministic`](https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/blob/5-1-stable/lib/arel/visitors/sqlserver.rb#L170-L179) at a lower level of the query builder.

The problem is that at that point, we don't have any information about the model, its just database and queries and we cannot get the PK information from your model so it tries to fetch this information on [`primary_Key_From_Table`](https://github.com/rails-sqlserver/activerecord-sqlserver-adapter/blob/5-1-stable/lib/arel/visitors/sqlserver.rb#L203-L208) and because the PK on your database is different from what you actually have on your model AND it is a composite PK, you get that error.

In summary, short term solution: change your table PK to reflect what is on your model.

Long term solution: we might need to investigate an alternative for when we have a situation like this. Maybe use the `IDENTITY` column if present and fallback to the PK and If PK is composite, choose only one of the columns, I don't know really what is the best solution.

**Comment by bwbasheer** [comment]
Interesting response.  Thank you.  One comment: This application has been upgraded from Rails 4.x, (and before that 3.x, where it worked fine.  Might that help in your further analysis, assuming you will be looking further into the issue for your own reasons?

**Comment by wpolicarpo** [comment]
As far as I can tell you, the adapter was rewritten from 4.2 to 5.0 to support parameterized queries and that's why there's such differences.

I don't know if we can really find and good a final solution at the adapter level tbh. The best thing to do would be to fix your model/table.

**Comment by bwbasheer** [comment]
OK.  Thanks.

**Comment by bwbasheer** [comment]
Followed your advice.  Issue seems resolved.  Thank you very much.
